### DISCUSSION TITLE
Why does getInitialProps cause an infinite loop when implementing pagination ?

### DISCUSSION THREAD

**Discussion by omar-dulaimi** [vercel/next.js#14956] [Help]
Hello,
I have been working on implementing server side pagination with nextjs, and this is how I thought was the best to accomplish that:
1- at first page load, when isServer is true, I get the data and count and page from backend and send it as response to frontend.
2- when page loads and renders the array of elements, it calls the on page change handler of the paginator that pushes to the route the current page number.
3- this causes a call back, triggering the infinite loop.

What am I doing wrong here?
this is the package i'm  using https://www.npmjs.com/package/react-paginate

**Comment by illuminist** [comment]
You may interested in using "shallow" option when do routing.

However, the better solution is just letting client side to run `getInitialProps` to data of the next or any page.

Or better yet, just don't use `getInitialProps` for listing pages as it's wasting resource for rendering page in serverside. Instead, use [SWR](https://swr.vercel.app/) to fetch page list in client side, with this, the listing page can be statically optimized.

**Reply by omar-dulaimi** [reply]
Hello @illuminist 
This is the first time I ever hear about SWR. We switched to Nextjs for SEO purposes as pages get served on server. How would using SWR play in that aspect?

how would it affect our SEO purposes if you could explain please.

**Reply by illuminist** [reply]
I used SWR for listing page only, for the reason, it doesn't need to be render from server and google bot is smart enough to traverse client side rendered pages to get a list of detail page.

For detail page, I used `getInitialProps` as it's more likely to be shared so it needs to be rendered for people to share the link with info card of the page.

**Reply by omar-dulaimi** [reply]
But the load time is important in my case, the listing pages play a big role in our website. So we used SSR to prepare all required data on the server, to decrease load time and improve our Web Vitals score since Google search console made this change this year.

Also, shallow routing is a great idea. I currently use this snippet to change pages for pagination:
```
    this.props.router.push({
             pathname: currentPath,
             query: currentQuery,
     });
```
But couldn't figure where to put the shallow option, it doesn't seem to be included in this object. The props come from withRouter.
All examples and docs I found used this method instead:
       ` Router.push('/post/[pid]?hello=123', '/post/abc?hello=123',  {
                shallow: true
            })`

So I did the same thing manually, but I get 404s:

`http://localhost:3000/_next/static/development/pages/search/%5Btype%5D/%5Bcat%5D/%5Barea%5D.js`
decoded:
`"http://localhost:3000/_next/static/development/pages/search/[type]/[cat]/[area].js"`

I tried using :type instead of [type] but it also didn't work.


This is how it's configured on the server:
```
                if ('/search/:type/:cat/:area' === route.path) {
                    return app.render(req, res, route.page);
                }
```

Do you have any Idea what might be the problem?

**Reply by illuminist** [reply]
It seems you appear to be using useRouter from react-router? NextJS has it own useRouter so use them accordingly.

Reading from your use case, if SSR is more preferred, then `getServerSideProps` will probably suite you the most.

**Reply by omar-dulaimi** [reply]
Not really, I'm using the withRouter wrapper from next/router package:
`import Router, { withRouter } from 'next/router';
`

I tried other ways as well, like:
```
        this.props.router.push(
            `/search/[type]/[cat]/[area]?${stringifyQs(currentQuery)}`,
            {
                pathname: currentPath,
                query: currentQuery,
            },
            { shallow: true }
        );

```

and

```
     this.props.router.push({
            pathname: currentPath,
            query: currentQuery
          }, 
          undefined, { shallow: true }
          )
```

and even tried making changes to the server as noted by a contributor here:
```
server.get('/search/:type/:cat/:area', (req, res) => {
         console.log("reached here...");   // this gets logged
        return app.render(
            req,
            res,
            '/search/[type]/[cat]/[area]',
            req.params
        );
});
```

All of them didn't work.

I read on SO that these params are considered page changes so the page will reload! 
https://stackoverflow.com/a/62845110/7186666

Although it doesn't make sense, why would a contributor give an answer that would never work?

**Reply by illuminist** [reply]
Router.push has call signature: `Router.push(href, as, options)` ,

for example
```jsx
Router.push('/post/[postId]/[date]', '/post/postnumber1/20200513?showComment=15', {shallow: true})
```

Your usage from the your original post is matched with react-router's Router, so that's why I thought there was some confusion here.
